File: build-wifakey-core/wifakey-core/python-reference/biohashing.py
import numpy as np
import hashlib
import hmac


def _hmac_gaussian(secret: bytes, row: int, col: int) -> float:
    """Tạo một số Gaussian từ HMAC(secret, row || col)."""
    msg = row.to_bytes(4, "big") + col.to_bytes(4, "big")
    digest = hmac.new(secret, msg, hashlib.sha256).digest()
    # Chuyển 4 byte đầu thành số nguyên, rồi dùng Box‑Muller
    u = int.from_bytes(digest[:4], "big") / 0xFFFFFFFF
    v = int.from_bytes(digest[4:8], "big") / 0xFFFFFFFF
    r = (-2.0 * np.log(u + 1e-15)) ** 0.5
    theta = 2.0 * np.pi * v
    return r * np.cos(theta)  # dùng 1 trong 2 giá trị độc lập


def generate_projection_matrix(user_secret: bytes, dim: int = 512) -> np.ndarray:
    M = np.zeros((dim, dim), dtype=np.float64)
    for i in range(dim):
        for j in range(dim):
            # Mỗi cặp (i,j) dùng một index để tạo giá trị Gaussian
            idx = i * dim + j
            # Để tạo 2 giá trị độc lập từ một lần gọi, ta tách chẵn/lẻ
            val = _hmac_gaussian(user_secret, idx, 0)
            M[i, j] = val
        # Chuẩn hoá hàng
        norm = np.linalg.norm(M[i, :])
        if norm > 0:
            M[i, :] /= norm
    return M.astype(np.float32)


def biohash_project(embedding: np.ndarray, user_secret: bytes) -> np.ndarray:
    M = generate_projection_matrix(user_secret, len(embedding))
    v_proj = np.dot(embedding, M.T)
    norm = np.linalg.norm(v_proj)
    if norm > 0:
        v_proj /= norm
    return v_proj.astype(np.float32)

File: build-wifakey-core/wifakey-core/src/biohashing.rs
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

/// Tạo một giá trị Gaussian từ HMAC(secret, row || col)
fn hmac_gaussian(secret: &[u8], row: usize, col: usize) -> f64 {
    let mut mac = HmacSha256::new_from_slice(secret).expect("HMAC key");
    mac.update(&row.to_be_bytes());
    mac.update(&col.to_be_bytes());
    let digest = mac.finalize().into_bytes();
    
    let u = u32::from_be_bytes(digest[..4].try_into().unwrap()) as f64 / 0xFFFFFFFFu64 as f64;
    let v = u32::from_be_bytes(digest[4..8].try_into().unwrap()) as f64 / 0xFFFFFFFFu64 as f64;
    
    let r = (-2.0f64 * (u + 1e-15).ln()).sqrt();
    let theta = 2.0 * std::f64::consts::PI * v;
    r * theta.cos()
}

pub fn generate_projection_matrix(user_secret: &[u8], dim: usize) -> Vec<f64> {
    let n = dim * dim;
    let mut m = vec![0.0f64; n];
    
    for i in 0..dim {
        for j in 0..dim {
            m[i * dim + j] = hmac_gaussian(user_secret, i, j);
        }
        // Chuẩn hoá hàng
        let start = i * dim;
        let mut norm_sq = 0.0;
        for j in 0..dim {
            norm_sq += m[start + j] * m[start + j];
        }
        let norm = norm_sq.sqrt();
        if norm > 0.0 {
            for j in 0..dim {
                m[start + j] /= norm;
            }
        }
    }
    m
}

pub fn biohash_project(embedding: &[f64], user_secret: &[u8]) -> Vec<f64> {
    let dim = embedding.len();
    let proj_matrix = generate_projection_matrix(user_secret, dim);
    let mut v_proj = vec![0.0f64; dim];
    for i in 0..dim {
        let mut sum = 0.0;
        for j in 0..dim {
            sum += proj_matrix[i * dim + j] * embedding[j];
        }
        v_proj[i] = sum;
    }
    let norm: f64 = v_proj.iter().map(|x| x * x).sum::<f64>().sqrt();
    if norm > 0.0 {
        v_proj.iter_mut().for_each(|x| *x /= norm);
    }
    v_proj
}